### ISSUE TITLE
Chapter 2, p 37 Instances: nominalization (or agglutination)

### ISSUE THREAD

**Issue by sgillesp-ai** [KGConf/Bookclub-DemystifyingOWL#11]
This question is regarding the instance _JohnDoeDoctorVisit12Jan

There is a lot of information packed in this individual instance of PatientVisit: 
- Information that looks like a data type property string: First name, last name, date
- Information of the type of careProvider: Doctor

Question 1: Is this bending the rules of OWL to insert data and subClasses into the name of an individual instance or is this a standard way of doing things? It's like some of the very long German words created through agglutination: Rindfleisch­etikettierungs­überwachungs­aufgaben­übertragungs­gesetz at 63 letters means "law delegating beef label monitoring" --thank you Wikipedia. 

Question 2: If it's standard, then there must be some form of automation in creating these strings in the real-life system. (I'm not sure I would want to type these out for every patient visit). Can you comment on how these are created? 
 
Apologies for late posting.

**Comment by uscholdm** [comment]
Yes, there is a lot of information packed into this IRI: `:_JohnDoeDoctorVisit12Jan`.  I made this choice for illustrative purposes for this book. I wanted something that one could guess the meaning from at a glance.  Coming up with unique IDs to use for IRIs in such cases is not easy.   

You have to first decide what combination of information will guarantee something that is unique.  There could easily be more than one person named John Doe who went to the doctor on 12Jan. There is not even a year that is indicated. In a recent project we faced this exact question.   In that case, the client already had unique ids for each patient and it was assumed that one patient could only be in one patient visit at a time, so the IRI was made by appending the patient ID to the date time of the visit.
For example, `:_PatientVisit_2439646_20200429_082200` is used to refer to the patient visit where the patient ID id is 2439646 and the date time is April 4, 2020 at 08:22am.  

We faced a bigger challenge in creating IRIs for measurements. For example:
- John Doe was weighed on April 4, 2020 at 10:32am

After careful consideration, we decided on IRIs that look like this: `:_Measurement_2439646_Weight_125lb_20200429_103200`. They are build up from the following information and generated automatically when creating triples
- Patient identifier (2439646)
- Name of characteristic being measured (Weight)
- The measured weight
  - number (125)
  - unit (lb)

An alternative to building up long meaningful IRIs is to use a hashing function.  They are harder for the human to read.

**Comment by dallemang** [comment]
This problem isn't unique to OWL (or the Semantic Web).  If you have a record in a relational database with all this information, and you want to refer to it (e.g., the appointment for this visit points to the visit, once it has happened), you'll need a way to identify it.  RDBMS's call this a "primary key", and typically they are assigned sequentially (and if it shows up where a person is going to read it, it is called something like a "visit ID").  Mike points out that hashes are harder for the human to read (or even impossible), but the same is true for visit ID's; you have no idea which care provider, patient or procedure was performed based just on a primary key in a database.

**Comment by uscholdm** [comment]
Dean makes good points.  When a client has taken care to create such primary keys, and they are used consistently throughout all tables, we often use them to create IRIs.  That is exactly what we did for the patient in the example above we used the primary key, 2439646, as part of the IRI.  That key might be used elsewhere for some other thing, so we prefix it with the class to ensure uniqueness, as in `:_Patient_2439646`.
